crates/amaru-ledger/src/rules/context.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2025 PRAGMA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use amaru_kernel::{
+    Anchor, CertificatePointer, DRep, Lovelace, PoolId, PoolParams, StakeCredential,
+    TransactionInput, TransactionOutput,
+};
+
+pub use simple::*;
+pub(crate) mod fake;
+mod simple;
+
+/// The BlockValidationContext is a collection of slices needed to validate a block
+pub trait BlockValidationContext:
+    PotsSlice + UtxoSlice + PoolsSlice + AccountsSlice + DRepsSlice
+{
+}
+
+/// The BlockPreparationContext is a collection of interfaces needed to prepare a block
+pub trait BlockPreparationContext<'a>:
+    PrepareUtxoSlice<'a> + PreparePoolsSlice<'a> + PrepareAccountsSlice<'a> + PrepareDRepsSlice<'a>
+{
+}
+
+/// An interface for interacting with the protocol pots.
+pub trait PotsSlice {
+    fn add_fees(&mut self);
+}
+
+// An interface for interacting with a subset of the UTxO state.
+pub trait UtxoSlice {
+    fn lookup(&self, input: &TransactionInput) -> Option<&TransactionOutput>;
+    fn consume(&mut self, input: &TransactionInput);
+    fn produce(&mut self, input: TransactionInput, output: TransactionOutput);
+}
+
+/// An interface to help constructing the concrete UtxoSlice ahead of time.
+pub trait PrepareUtxoSlice<'a> {
+    fn require_input(&'_ mut self, input: &'a TransactionInput);
+}
```

Comment:
This is a matter of taste and opinion: I only tend to split up traits when the individual pieces shall be used independently from each other. This PR so far doesn’t seem to show such independence, instead using the preparation context trait.

Reply:
Good point, though the PR mostly introduces some preparatory work, but the traits are indeed used independently!